A11y: Fix toggletip predictable focus for keyboard users

ckbedwell opened the pull request:
**What is this feature?**
This is an accessibility enhancement for the `<ToggleTip />` component so that focus order is predictable and expected as per [WCAG 2.1 Success Criterion 2.4.3 - Focus Order](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html).

The implementation is following Adrian Roselli's disclosure [widgets post](https://adrianroselli.com/2020/05/disclosure-widgets.html), and specifically his [toggle-tip example](https://codepen.io/aardrian/pen/NWpoVQd). TPGi have a [similar implementation](https://www.tpgi.com/simple-standalone-toggletip-widget-pattern/). _I've put a note in the additional discussion about toggle-tip implementations._

**Why do we need this feature?**
So that keyboard users and users of assistive technologies can easily and predictably interact with toggletip components.

**Who is this feature for?**
Everybody using the Toggletip component!

**Which issue(s) does this PR fix?**:
Un-filed issue (being proactive and trying to fix it instead 😄 ).

**Special notes for your reviewer:**
_Apologies for how long this is about to get..._

1. One of the main aspects of this PR is I've removed the `<Portal />` implementation as this is the root problem when it comes to popups, Focus order and assistive technologies. I've utilised `react-popper-tooltip`'s `strategy: fixed` [option](https://popper.js.org/docs/v2/constructors/#strategy). I've tested the implementation on the page, inside a Modal and a Drawer and can see no discernible visual or performative difference between this fixed implementation vs the Portal version . I've tested on the following devices:

**MacOS [13.4.1 (c) (22F770820d)]**
- Chrome 114.0.5735.198 & 115.0.5790.102
- Firefox 115.0.2 (64-bit)
- Safari 16.5.2
_All checked with keyboard-only, mouse-only and with Voiceover enabled_

**Windows 11 [22H2]**
- Chrome 114.0.5735.199 & 115.0.5790.102
- Firefox 114.0.1
- Edge 114.0.1823.86
_All checked with keyboard-only, touch-only, mouse-only and with NVDA enabled_

**Android [13 TKQ1.22829.002]**
- Chrome 114.0.5735.196
- Firefox 115.2.1
_All checked with touch interface. (I'm no good with TalkBack on my phone yet 😞)_

2. I've put in a temporary Storybook story (which I'll remove afterwards) with the basic and multi input Toggletips in a `<Modal />` and in a `<Drawer />` and you can even open a Modal in the Drawer version to see the Toggletips still appearing correctly (there's a separate issue with click outside handlers with mouse but it works well with Keyboard!) to make sure the strategy fixed works as intended and there are no regressions. Please check any other components where there is concern in replacing a Portal with the fixed strategy.

3. At the moment I've limited the scope of this PR to fix the immediate Focus issue but I think there is a wider conversation to be had about Toggletip implementations and recommendations. I've based this Focus approach on the Toggletip on Adrian Roselli's accessible disclosure approach as that's what seemed to be the intent (and I believe to be one of the better implementations) but there are details of toggletips which differ between different libraries. 
 
   * If we take [Inclusive Component's approach](https://inclusive-components.design/tooltips-toggletips/) to Toggletips, they automatically close after a keyboard user tabs out of the toggle information. I think there is an argument for this behaviour in _some_ but not all cases. I've left a convoluted example in the temporary Storybook where it would be useful for a Toggletip to remain open whilst reading / interacting with other information (obviously we would want to recommend different design solutions to problems so take this example with a pinch of salt). To extend on this as an option, I don't think the `closeOnOutsideClick` should always be active for the popper as there are users of assistive technologies who would benefit without it. (touch-screen users, users with screen zoom, users with varying cognitive disabilities, users who highlight text as they read, etc.).
   
   * I think there is a use case that the Focus should be trapped within the toggletip at times. The multi-element Toggletip story is a candidate for this use case. However, creating a toggletip in this way is essentially creating a new version of a modal as they would have very similar functionality at this point (especially to users of assistive technologies).
   
   * With the two points above, I would propose two extra props added to handle these use-cases so the developer can decide when to use them as appropriate.

4. I originally considered using a mixture of Portal / FocusScope / autoFocus, however this would fail the test I've added for keyboard users who are moving their focus backwards and would be a WCAG failure.

5. Very last note on accessibility in toggletips: If a developer ~~chooses to turn off the `closeButton` prop and~~ _(edit: discovered it doesn't matter if this is enabled or not as it is outside the content block)_ has no interactive content within the toggletip, they are potentially limiting users of assistive technologies' ability to read / interact with the content. I think it would be good to add to the toggletip documentation that they might have to add `tabIndex=0` to their content if it is sufficiently long.

Please check that:
- [x] It works as expected across the various components that might have a toggletip in them.
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

ckbedwell committed:
Updated toggletip to use strategy fixed, added tests and temporary story so it can easily be checked

ckbedwell committed:
Added FocusScope restoreFocus and appropriate tests to toggletip

ckbedwell committed:
Open toggletip in test for making sure focus remains when using escape

ckbedwell committed:
Add aria-expanded to toggletip toggle child

ckbedwell committed:
Merge branch 'main' into ckbedwell/toggletip-focusscope

ckbedwell committed:
Added to temp story for Toggletip

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=72100) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=72100) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=72100) it.</sub>

ashharrison90 commented on packages/grafana-ui/src/components/Toggletip/ToggletipTemp.story.tsx:
assuming this was your test environment, think we should either:
- remove this
- rename it (to not be temporary) and clean it up a bit

ckbedwell replied:
Yeah, that temp story is very messy 😂 -- I wanted to make it as easy as possible for reviewers so they could see this approach in action rather than them having to create their own Modal / Drawer components to check it.

The intent was to delete it before it gets merged if the rest of the PR is acceptable but there might be some things in there that could be salvaged and added to 'real' Storybook stories before that happens 😃

ashharrison90 reviewed (approved):
this looks fantastic! 🙌 

left a comment about the temporary story, also pinging @JoaoSilvaGrafana @yduartep for 👀

JoaoSilvaGrafana commented:
Great job on this! 🎉 👏  And thanks for raising those other points as well. Especially about point 5, feel free to add some documentation either to storybook or to the [Saga doc](https://github.com/grafana/design-system/blob/main/docs/05-Components/toggletip.mdx) about this 🙌 
About the 2 new props, I think maybe we would prefer to wait for some specific use cases within Grafana that would need it before we add it, just weary of adding too many props to components like we have in the past 😅 

One thing I noticed on running this that seems to be different is that there seems to be one more tab stop in the ToggleTip 🤔 
For example on the "Basic" story:
- Tab onto the button
- Press enter to open Toggletip
- Tab onto the "x" button
- Shift-tab backwards, the focus is somewhere else
- Shift-tab backwards again, focus is now on the button that opens the toggletip

Do you know why this might be?

ckbedwell committed:
Merge branch 'main' into ckbedwell/toggletip-focusscope

ckbedwell committed:
Remove focusScope for toggletip and handle focus restoration manually

ckbedwell committed:
Remove toggletip temp story and add toggletip long content story

ckbedwell commented:
> Especially about point 5, feel free to add some documentation either to storybook or to the [Saga doc](https://github.com/grafana/design-system/blob/main/docs/05-Components/toggletip.mdx) about this

I've removed the temporary story now and I've added a story about long content specifically. I'll take a look at putting a  separate PR up in the Saga documentation 😃 

> One thing I noticed on running this that seems to be different is that there seems to be one more tab stop in the ToggleTip
> ...
> Do you know why this might be?

Good catch! Annoyingly it was a bug caused by the `<FocusScope restoreFocus>` from `react-aria`. I've removed it completely and added the focus restoration to the `closeToggletip` function. I've updated the expected focus test to ensure this doesn't regress in the future.

ashharrison90 commented:
@ckbedwell should this have been closed? 😱

ckbedwell commented:
> @ckbedwell should this have been closed? 😱

Nope 😂 No idea how I managed that. Thanks for noticing!

JoaoSilvaGrafana reviewed (approved):
Lgtm now!!

joshhunt committed:
Merge remote-tracking branch 'origin/main' into ckbedwell/toggletip-focusscope
